lib/spack/spack/database.py

```diff
@@ -0,0 +1,290 @@
+        except MarkedYAMLError, e:
+            raise SpackYAMLError("error parsing YAML database:", str(e))
+
+        if file==None:
+            return
+
+        data = {}
+        for index, sp in file['database'].items():
+            spec = Spec.from_node_dict(sp['spec'])
+            deps = sp['dependency_indices']
+            path = sp['path']
+            dep_hash = sp['hash']
+            db_entry = {'deps':deps, 'spec': spec, 'path': path, 'hash':dep_hash}
+            data[index] = db_entry
+
+        for sph in data.values():
+            for idx in sph['deps']:
+                sph['spec'].dependencies[data[idx]['spec'].name] = data[idx]['spec']
+
+        self.data = data.values()
+
+
+    def read_database(self):
+        """
+        Re-read Database from the data in the set location
+        If the cache is fresh, return immediately.
+        Implemented with mkdir locking for the database file.
+        """
+        if not self.is_dirty():
+            return
+
+        lock=0
+        while lock==0:
+            try:
+                os.mkdir(self.lock_path)
+                lock=1
+            except OSError as err:
+                pass
+
+        #The try statement ensures that a failure won't leave the
+        #database locked to other processes.
+        try:
+            if os.path.isfile(self.file_path):
+                with open(self.file_path,'r') as f:
+                    self.from_yaml(f)
+            else:
+            #The file doesn't exist, construct empty data.
+                self.data = []
+        except:
+            os.rmdir(self.lock_path)
+            raise
+
+        os.rmdir(self.lock_path)
+
+
+    def write_database_to_yaml(self,stream):
+        """
+        Replace each spec with its dict-node form
+        Then stream all data to YAML
+        """
+        node_list = []
+        spec_list = [sph['spec'] for sph in self.data]
+
+        for sph in self.data:
+            node = {}
+            deps = []
+            for name,spec in sph['spec'].dependencies.items():
+                deps.append(spec_list.index(spec))
+            node['spec']=Spec.to_node_dict(sph['spec'])
+            node['hash']=sph['hash']
+            node['path']=sph['path']
+            node['dependency_indices']=deps
+            node_list.append(node)
+
+        node_dict = dict(enumerate(node_list))
+        return yaml.dump({ 'database' : node_dict},
+                         stream=stream, default_flow_style=False)
+
+
+    def write(self):
+        """
+        Write the database to the standard location
+        Implements mkdir locking for the database file
+        """
+        lock=0
+        while lock==0:
+            try:
+                os.mkdir(self.lock_path)
+                lock=1
+            except OSError as err:
+                pass
+
+        #The try statement ensures that a failure won't leave the
+        #database locked to other processes.
+        try:
```

Comment:
So, if you acquire the lock on the DB after someone else wrote to it, then immediately write, you'll blow away their changes and you'll lose the earlier of two concurrently installed packages.  `write` should probably check for `is_dirty` and re-read the DB at least, THEN merge new stuff.  We should probably discuss how to handle concurrent modifications, as there may be more complications we haven't thought of.

The DB should probably be checked before installing each dependency package, and Spack should skip the dep and use what's there if it finds that some other instance has installed a package.

I think what you want is atomic installs of each package, and use-after-install consistency among concurrent spack instances.  I know at least NFS should support that (right, @mplegendre?  NFS has open-after-close consistency).  Not sure about other filesystems but it should be doable.  I just want to avoid a situation where we're installing into an already-installed package directory, and one build uses a partially written file from the second install.  One way to safeguard against that would be to make the install dirs read-only, something other pkg managers (Nix, Guix, Hashdist) already do.